### ISSUE TITLE
Cannot read property 'canonicalUrl' of undefined

### ISSUE THREAD

**Issue by ryanflowers** [angular/router#272]
As of v0.5.1 I am seeing an unhandled exception of "Cannot read property 'canonicalUrl' of undefined". 

This occurs when changing routes. I have created a plunker(http://plnkr.co/edit/4vczoQv6Ufm5QUkx3azB?p=preview) to demonstrate the issue. The plunker does a redirect to a 'test1' component when the plunker loads. As you can see when changing routes the following code is run within the router.

forEach(lastHandler.components, (function(componentName, viewportName) {
          instruction.viewports[viewportName] = $__0.recognize(childUrl, componentName);
        }));
        instruction.canonicalUrl += instruction.viewports[Object.keys(instruction.viewports)[0]].canonicalUrl;

From what I can see the lastHandler.components are the components from the previous route. The code iterates them and ends up with undefined viewports for the current instruction if they are not recognized for the new route. Thus instruction.viewports[Object.keys(instruction.viewports)[0]] throws the exception.

**Comment by robgrimball** [comment]
Just switched to 0.5.3 from the last commit that still allowed $scope to be injected into the controllers (a6166ffb6794e64669ddd1c91dfcfc7f49d9614a) and am running into this same issue.  Exactly the same as described by ryanflowers.

**Comment by ljungbla** [comment]
@ryanflowers  This seems to happen because of the empty path. If you add a '/' to it then is does not throw anymore. (And also if the MainController is added and the ng-viewport="main" is changed to 'div ng-viewport="main"')

http://plnkr.co/edit/A3YxOobjU8EdvbQ3cqAP?p=preview

But this error is thrown when I insert a route that doesn't exist in my app. What normally should be handeld as a 404.

**Comment by ryanflowers** [comment]
Yes in debugging it the code seemed to think it was a path not declared in my routes. The strange part is that when I remove the hashtag from my urls they work again. Does the new router no longer support hashtags for some reason? Anyone else run into this? Could just be my setup..

**Comment by stakach** [comment]
This also happens when clicking a `mailto:` link. Doesn't prevent it working, just throws this error.

**Comment by cooervo** [comment]
As @ryanflowers said I removed `/#` from beginning of link and it fixed the issue.
before `/#/foobar`
After: `/foobar`

I don't get the canonicalUrl error anymore.
